### PR TITLE
Merge ResourceUpdate and Transaction

### LINKED ISSUE #2322: Merge ResourceUpdate and Transaction.
`ResourceUpdate` was a first step in the direction of bundling several operations in a single API message, and `Transaction` offers that as well with in a more powerful way. We can just move `ResourceUpdates` methods directly in `Transaction` to simplify the API a bit.

### PR DESCRIPTION
**Pull request by hgallagher1993**
Fixes #2322

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/2718)

### CODE REVIEW

File: webrender_api/src/api.rs [hunk 1]
```diff
@@ -357,14 +263,94 @@ impl Transaction {
...
+                variations,
+            }));
+    }
+
+    pub fn delete_font_instance(&mut self, key: FontInstanceKey) {
+        self.resource_updates.push(ResourceUpdate::DeleteFontInstance(key));
+    }
+
+    pub fn merge(&mut self, mut other: Vec<ResourceUpdate>) {
+        self.resource_updates.append(&mut other);
```
Comment (kvark) [inline webrender_api/src/api.rs]: nit: could be `extend(other)`

File: webrender_api/src/api.rs [hunk 2]
```diff
@@ -357,14 +263,94 @@ impl Transaction {
...
+                platform_options,
+                variations,
+            }));
+    }
+
+    pub fn delete_font_instance(&mut self, key: FontInstanceKey) {
+        self.resource_updates.push(ResourceUpdate::DeleteFontInstance(key));
+    }
+
+    pub fn merge(&mut self, mut other: Vec<ResourceUpdate>) {
```
Comment (kvark) [inline webrender_api/src/api.rs]: does this method make sense for the user to call? if yes, needs to be renamed for clarity. If no, needs to be hidden (do we need it at all then?)

**Comment by hgallagher1993** [comment]
@kvark Ya the CI was a mistake I never did a `cargo build`, I just built Servo with my local Webrender so I'll try and get the problems with examples/ and wrench/ sorted hopefully tomorrow or Friday after work

@staktrace I've literally never done anything with Gecko / Firefox apart from download the source code and read a bit of the developer guide, so I guess if the code changes are simple enough like it was to get Servo to compile I'll happily give it a go but if it's a bit more involved someone else might have to it

**Comment by staktrace** [comment]
@hgallagher1993 Ok, once you fix the standalone WR build I or somebody else can help you do the gecko side of stuff.

**Comment by hgallagher1993** [comment]
Apologies for the delay but anyway example/ and wrench/ should be compiling now

File: webrender_api/src/api.rs (hunk 2 above)
Comment (hgallagher1993) [reply webrender_api/src/api.rs]: Every instance of `txn.update_resources(resources);` (so by extension `merge()` as well) I ended up deleting so is it really needed anymore?

**Comment by staktrace** [comment]
Looks like it's still failing the taskcluster CI checks. Raw log is at https://taskcluster-artifacts.net/bArGBUPWQ1WO7NfjzJ5xCA/0/public/logs/live_backing.log - looks like lint failures

```
  Progress: 21% (162/737)
./wrench/src/yaml_frame_writer.rs:506: use &[T] instead of &Vec<T>
  Progress: 22% (163/737)
  Progress: 22% (164/737)
./wrench/src/ron_frame_writer.rs:90: use &[T] instead of &Vec<T>
```

(among others)

**Comment by hgallagher1993** [comment]
@staktrace those few tidy errors should be fixed now (and an unused import), but is there anyway to do that check locally?

**Comment by staktrace** [comment]
You should be able to check it locally by running the following commands in your webrender checkout:

```
virtualenv --python=python2.7 ../stidy-venv # create a virtualenv for python
source ../stidy-venv/bin/activate # activate virtualenv
pip install PyYAML voluptuous servo-tidy # install the tool
servo-tidy # run the tool
deactivate # deactivate virtualenv
```

**Comment by staktrace** [comment]
Looks like there's still an error on AppVeyor.

**Comment by hgallagher1993** [comment]
Right fingers crossed with this try :laughing:

**Comment by staktrace** [comment]
I wrote the gecko-side patch and pushed it to try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=a96ef6f8a83d638784062da147c95a6d47a52251

Assuming that's green this should be good to go.

**Pull request merged**